### PR TITLE
docs: annotate io mod with doc_cfg

### PR DESCRIPTION
**Pull request by carllerche**
Annotates types in `tokio::io` module with their required feature flag.
This annotation is included in generated documentation.
    
Notes:
    
* The annotation must be on the type or function itself. Annotating just
   the re-export is not sufficient.

* The annotation must be **inside** the `pin_project!` macro or it is
   lost.

### CODE REVIEW

**Comment by carllerche** [comment]
Once we land on specifics for style, it can be applied to the rest of the crate.

File: tokio/src/macros/cfg.rs [hunk 1]
```diff
@@ -66,25 +80,40 @@ macro_rules! cfg_io_blocking {
 
 macro_rules! cfg_io_driver {
     ($($item:item)*) => {
-        $( #[cfg(feature = "io-driver")] $item )*
+        $(
+            #[cfg(feature = "io-driver")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "io-util")))]
```
Comment (taiki-e) [inline tokio/src/macros/cfg.rs]:
```suggestion
            #[cfg_attr(docsrs, doc(cfg(feature = "io-driver")))]
```

File: tokio/src/io/util/flush.rs [hunk 2]
```diff
@@ -5,12 +5,15 @@ use std::io;
...
-/// Created by the [`AsyncWriteExt::flush`] function.
-#[derive(Debug)]
-pub struct Flush<'a, A: ?Sized> {
-    a: &'a mut A,
+cfg_io_util! {
+    /// A future used to fully flush an I/O object.
+    ///
+    /// Created by the [`AsyncWriteExt::flush`] function.
+    #[derive(Debug)]
+    #[cfg_attr(docsrs, doc(cfg(feature = "io-util")))]
```
Comment (taiki-e) [inline tokio/src/io/util/flush.rs]:
```suggestion
```

**Comment by taiki-e** [comment]
Perhaps we need to update CI config.

https://github.com/tokio-rs/tokio/blob/8546ff826db8dba1e39b4119ad909fb6cab2492a/ci/azure-deploy-docs.yml#L18

https://github.com/tokio-rs/tokio/blob/8546ff826db8dba1e39b4119ad909fb6cab2492a/azure-pipelines.yml#L77

File: tokio/src/io/util/read_to_end.rs [hunk 3]
```diff
@@ -6,12 +6,14 @@ use std::mem::MaybeUninit;
...
-#[derive(Debug)]
-#[must_use = "futures do nothing unless you `.await` or poll them"]
-pub struct ReadToEnd<'a, R: ?Sized> {
-    reader: &'a mut R,
-    buf: &'a mut Vec<u8>,
-    start_len: usize,
+cfg_io_util! {
+    #[derive(Debug)]
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    pub struct ReadToEnd<'a, R: ?Sized> {
```
Comment (taiki-e) [inline tokio/src/io/util/read_to_end.rs]: As `ReadToEnd` is used internally even when `io-util` feature is disabled, it seems to be necessary to use an attribute.

Comment (carllerche) [reply tokio/src/io/util/read_to_end.rs]: `ReadToEnd` is only **publicly* available w/ the feature. I do need to fix it though.

**Pull request merged**